app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,20 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    # You need to implement the method below in your model (e.g. app/models/user.rb)
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      self.current_user = @user
+      sign_in_and_redirect @user, :event => :authentication #this will throw if @user is not activated
+      set_flash_message(:notice, :success, :name => @user.name) if is_navigational_format?
+
+    else
+      session["devise.facebook_data"] = request.env["omniauth.auth"]
```

Comment:
I think this assignment here is not really needed. I see that you used the code from https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview, but I think we can slim it down a bit :)